sdk/storage/Azure.Storage.Blobs/src/PartitionedDownloader.cs

```diff
@@ -0,0 +1,264 @@
+                // stream when each segment finishes.
+                Queue<Task<Response<BlobDownloadInfo>>> runningTasks = new Queue<Task<Response<BlobDownloadInfo>>>();
+                runningTasks.Enqueue(initialResponseTask);
+
+                // Fill the queue with tasks to download each of the remaining
+                // ranges in the blob
+                foreach (HttpRange httpRange in GetRanges(initialLength, totalLength))
+                {
+                    // Add the next Task (which will start the download but
+                    // return before it's completed downloading)
+                    runningTasks.Enqueue(_client.DownloadAsync(
+                        httpRange,
+                        conditionsWithEtag,
+                        rangeGetContentHash: false,
+                        cancellationToken));
+
+                    // If we have fewer tasks than alotted workers, then just
+                    // continue adding tasks until we have _maxWorkerCount
+                    // running in parallel
+                    if (runningTasks.Count < _maxWorkerCount)
+                    {
+                        continue;
+                    }
+
+                    // Once all the workers are busy, wait for the first
+                    // segment to finish downloading before we create more work
+                    await ConsumeQueuedTask().ConfigureAwait(false);
+                }
+
+                // Wait for all of the remaining segments to download
+                while (runningTasks.Count > 0)
+                {
+                    await ConsumeQueuedTask().ConfigureAwait(false);
+                }
+
+                return initialResponse.GetRawResponse();
+
+                // Wait for the first segment in the queue of tasks to complete
+                // downloading and copy it to the destination stream
+                async Task ConsumeQueuedTask()
+                {
+                    // Don't need to worry about 304s here because the ETag
+                    // condition will turn into a 412 and throw a proper
+                    // RequestFailedException
+                    using BlobDownloadInfo result = await runningTasks.Dequeue().ConfigureAwait(false);
+
+                    // Even though the BlobDownloadInfo is returned immediately,
+                    // CopyToAsync causes ConsumeQueuedTask to wait until the
+                    // download is complete
+                    await CopyToAsync(result, destination, cancellationToken).ConfigureAwait(false);
+                }
+            }
+            catch (Exception ex)
+            {
+                scope.Failed(ex);
+                throw;
+            }
+            finally
+            {
+                scope.Dispose();
+            }
+        }
+
+        public Response DownloadTo(Stream destination, BlobRequestConditions conditions, CancellationToken cancellationToken)
```

Comment:
Parallel download would end up doing sync-over-async which we don't want.  This will instead download in a sequence of range calls.  It's here more for symmetry than exposing an essential feature.